File: src/bigclaw/cost_control.py
from dataclasses import dataclass
from typing import Dict, Optional

from .models import Task


@dataclass
class BudgetDecision:
    status: str
    estimated_cost: float
    remaining_budget: float
    reason: str


class CostController:
    def __init__(self, medium_hourly_costs: Optional[Dict[str, float]] = None):
        self.medium_hourly_costs = medium_hourly_costs or {
            "docker": 2.0,
            "browser": 4.0,
            "vm": 8.0,
            "none": 0.0,
        }

    def estimate_cost(self, medium: str, duration_minutes: int) -> float:
        hourly = self.medium_hourly_costs.get(medium, 0.0)
        return round(hourly * (max(0, duration_minutes) / 60.0), 2)

    def evaluate(self, task: Task, medium: str, duration_minutes: int, spent_so_far: float = 0.0) -> BudgetDecision:
        estimated = self.estimate_cost(medium, duration_minutes)
        effective_budget = float(task.budget + task.budget_override_amount)
        remaining = round(effective_budget - spent_so_far - estimated, 2)

        if effective_budget <= 0:
            return BudgetDecision(
                status="allow",
                estimated_cost=estimated,
                remaining_budget=remaining,
                reason="budget not set",
            )

        if remaining >= 0:
            return BudgetDecision(
                status="allow",
                estimated_cost=estimated,
                remaining_budget=remaining,
                reason="within budget",
            )

        downgraded_medium = "docker" if medium in {"browser", "vm"} else "none"
        if downgraded_medium != medium:
            downgraded_estimated = self.estimate_cost(downgraded_medium, duration_minutes)
            downgraded_remaining = round(effective_budget - spent_so_far - downgraded_estimated, 2)
            if downgraded_remaining >= 0:
                return BudgetDecision(
                    status="degrade",
                    estimated_cost=downgraded_estimated,
                    remaining_budget=downgraded_remaining,
                    reason=f"degrade to {downgraded_medium} to stay within budget",
                )

        return BudgetDecision(
            status="pause",
            estimated_cost=estimated,
            remaining_budget=remaining,
            reason="budget exceeded",
        )

File: bigclaw-go/internal/scheduler/cost_control.go
package scheduler

import (
	"fmt"
	"math"

	"bigclaw-go/internal/domain"
)

type BudgetDecision struct {
	Status          string
	EstimatedCost   float64
	RemainingBudget float64
	Reason          string
}

type CostController struct {
	mediumHourlyCosts map[string]float64
}

func NewCostController(mediumHourlyCosts map[string]float64) CostController {
	if mediumHourlyCosts == nil {
		mediumHourlyCosts = map[string]float64{
			"docker":  2.0,
			"browser": 4.0,
			"vm":      8.0,
			"none":    0.0,
		}
	}
	return CostController{mediumHourlyCosts: mediumHourlyCosts}
}

func (c CostController) EstimateCost(medium string, durationMinutes int) float64 {
	hourly := c.mediumHourlyCosts[medium]
	return round2(hourly * (math.Max(0, float64(durationMinutes)) / 60.0))
}

func (c CostController) Evaluate(task domain.Task, medium string, durationMinutes int, spentSoFar float64) BudgetDecision {
	estimated := c.EstimateCost(medium, durationMinutes)
	effectiveBudget := float64(task.BudgetCents)/100 + task.BudgetOverrideAmount
	remaining := round2(effectiveBudget - spentSoFar - estimated)

	if effectiveBudget <= 0 {
		return BudgetDecision{
			Status:          "allow",
			EstimatedCost:   estimated,
			RemainingBudget: remaining,
			Reason:          "budget not set",
		}
	}

	if remaining >= 0 {
		return BudgetDecision{
			Status:          "allow",
			EstimatedCost:   estimated,
			RemainingBudget: remaining,
			Reason:          "within budget",
		}
	}

	downgradedMedium := "none"
	if medium == "browser" || medium == "vm" {
		downgradedMedium = "docker"
	}
	if downgradedMedium != medium {
		downgradedEstimated := c.EstimateCost(downgradedMedium, durationMinutes)
		downgradedRemaining := round2(effectiveBudget - spentSoFar - downgradedEstimated)
		if downgradedRemaining >= 0 {
			return BudgetDecision{
				Status:          "degrade",
				EstimatedCost:   downgradedEstimated,
				RemainingBudget: downgradedRemaining,
				Reason:          fmt.Sprintf("degrade to %s to stay within budget", downgradedMedium),
			}
		}
	}

	return BudgetDecision{
		Status:          "pause",
		EstimatedCost:   estimated,
		RemainingBudget: remaining,
		Reason:          "budget exceeded",
	}
}

func round2(value float64) float64 {
	return math.Round(value*100) / 100
}
